gxexternalproviders/src/main/java/com/genexus/db/driver/ExternalProviderS3.java

```diff
@@ -61,38 +60,75 @@
 
-    private AmazonS3 client;
-    private String bucket;
-    private String folder;
-	private String region = DEFAULT_REGION;
+	private AmazonS3 client;
+	private String bucket;
+	private String folder;
+
 
     public ExternalProviderS3(String service) {
         this(Application.getGXServices().get(service));
     }
 
-    public ExternalProviderS3(GXService providerService) {
-		bucket = Encryption.decrypt64(providerService.getProperties().get(BUCKET)).toLowerCase();
-		folder = providerService.getProperties().get(FOLDER);
-		region = providerService.getProperties().get(REGION);
+	public ExternalProviderS3(String accessKey, String secretKey, String bucketName, String region, String folder, String endpoint) {
+		init(accessKey, secretKey, bucketName, region, folder, endpoint);
+	}
 
-    	AWSCredentials credentials = new BasicAWSCredentials(Encryption.decrypt64(providerService.getProperties().get(ACCESS_KEY_ID)), Encryption.decrypt64(providerService.getProperties().get(SECRET_ACCESS_KEY)));
-		client = AmazonS3ClientBuilder.standard().withCredentials(new AWSStaticCredentialsProvider(credentials)).withRegion(region).build();
+	public ExternalProviderS3(GXService providerService) {
+		String accessKey = Encryption.decrypt64(providerService.getProperties().get(ACCESS_KEY_ID));
+		String secretKey = Encryption.decrypt64(providerService.getProperties().get(SECRET_ACCESS_KEY));
 
-		setEndpoint(providerService.getProperties().get(ENDPOINT));
-        bucketExists();
-        ensureFolder(folder);
-    }
+		String bucket = Encryption.decrypt64(providerService.getProperties().get(BUCKET)).toLowerCase();
+		String folder = providerService.getProperties().get(FOLDER);
+		String region = providerService.getProperties().get(REGION);
+		String endpointValue = providerService.getProperties().get(STORAGE_ENDPOINT);
 
+		if (endpointValue.equals("custom")) {
+			endpointValue = providerService.getProperties().get(STORAGE_CUSTOM_ENDPOINT);;
+		}
+		init(accessKey, secretKey, bucket, region, folder, endpointValue);
+	}
+
+	private void init(String accessKey, String secretKey, String bucketName, String region, String folder, String endpoint) {
+		if (this.client == null) {
+			if (region.length() == 0) {
+				region = DEFAULT_REGION;
+			}
+			this.bucket = bucketName;
+			this.folder = folder;
+			this.client = buildS3Client(accessKey, secretKey, endpoint, region);
 
-	private void setEndpoint(String endpoint) {
-		if (endpoint.equals(ACCELERATED)) {
-			client.setS3ClientOptions(S3ClientOptions.builder().setAccelerateModeEnabled(true).build());
+			bucketExists();
+			ensureFolder(folder);
 		}
-		if (endpoint.equals(DUALSTACK)) {
-			client.setS3ClientOptions(S3ClientOptions.builder().enableDualstack().setAccelerateModeEnabled(true).build());
+	}
+
+	private AmazonS3 buildS3Client(String accessKey, String secretKey, String endpoint, String region) {
+		AmazonS3 s3Client;
+		AWSCredentials credentials = new BasicAWSCredentials(accessKey, secretKey);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If possible, use temporary security credentials (IAM roles) instead of long-term access keys.
Long-term access keys, such as those associated with IAM users and AWS account root users, remain valid until you manually revoke them. However, temporary security credentials obtained through IAM roles and other features of the AWS Security Token Service expire after a short period of time. Use temporary security credentials to help reduce your risk in case credentials are accidentally exposed.
Learn more about [best practices for managing AWS access keys](https://docs.aws.amazon.com/general/latest/gr/aws-access-keys-best-practices.html).